Windows: Visual C++ C4316

There is an alignment issue reported which could hit performance on Windows

..\src\xxhash3.cc(162): warning C4316: 'XXHash3': object allocated on the heap may not be aligned 64

CI reference: https://ci.appveyor.com/project/ktrongnhan/xxhash-addon/builds/29695486/job/tbw0t6ek5efa4uh6

--- a/src/xxhash3.cc
+++ b/src/xxhash3.cc
@@ -5,7 +5,8 @@
 XXHash3::XXHash3(XXH64_hash_t seed)
     : env_(nullptr), wrapper_(nullptr), seed_(seed)
 {
-  XXH3_64bits_reset_withSeed(&state_, seed);
+  state_ = XXH3_createState();
+  XXH3_64bits_reset_withSeed(state_, seed);
 }
 
 XXHash3::XXHash3(const void *secret, size_t secretSize)
@@ -19,7 +20,8 @@ XXHash3::XXHash3(const void *secret, size_t secretSize)
   assert(secret_);
 
   secret_ = memcpy(secret_, secret, secretSize);
-  XXH3_64bits_reset_withSecret(&state_, secret_, secretSize);
+  state_ = XXH3_createState();
+  XXH3_64bits_reset_withSecret(state_, secret_, secretSize);
 }
 
 XXHash3::~XXHash3()
@@ -28,6 +30,12 @@ XXHash3::~XXHash3()
   {
     free(secret_);
     secret_ = NULL;
+  }
+
+  if (state_ != NULL)
+  {
+    XXH3_freeState(state_);
+    state_ = NULL;
   }
 
   napi_delete_reference(env_, wrapper_);
@@ -269,7 +277,7 @@ napi_value XXHash3::Update(napi_env env, napi_callback_info info)
   status = napi_unwrap(env, jsthis, reinterpret_cast<void **>(&hasher));
   assert(status == napi_ok);
 
-  XXH_errorcode hash_error = XXH3_64bits_update(&(hasher->state_),
+  XXH_errorcode hash_error = XXH3_64bits_update(hasher->state_,
                                                 reinterpret_cast<void *>(data),
                                                 buf_len);
   assert(hash_error == XXH_OK);
@@ -290,7 +298,7 @@ napi_value XXHash3::Digest(napi_env env, napi_callback_info info)
   assert(status == napi_ok);
 
   XXH64_canonical_t canonical_sum;
-  XXH64_hash_t sum = XXH3_64bits_digest(&(hasher->state_));
+  XXH64_hash_t sum = XXH3_64bits_digest(hasher->state_);
 
   // canonicalize the hash value
   XXH64_canonicalFromHash(&canonical_sum, sum);
@@ -317,9 +325,9 @@ napi_value XXHash3::Reset(napi_env env, napi_callback_info info)
   assert(status == napi_ok);
 
   if (hasher->secret_ == NULL)
-    assert(XXH3_64bits_reset_withSeed(&(hasher->state_), hasher->seed_) == XXH_OK);
+    assert(XXH3_64bits_reset_withSeed(hasher->state_, hasher->seed_) == XXH_OK);
   else
-    assert(XXH3_64bits_reset_withSecret(&(hasher->state_), hasher->secret_, hasher->secretSize_) == XXH_OK);
+    assert(XXH3_64bits_reset_withSecret(hasher->state_, hasher->secret_, hasher->secretSize_) == XXH_OK);
 
   return nullptr;
 }

--- a/src/xxhash3.h
+++ b/src/xxhash3.h
@@ -33,7 +33,7 @@ private:
   napi_env env_;
   napi_ref wrapper_;
   XXH64_hash_t seed_;
-  XXH3_state_t state_;
+  XXH3_state_t* state_ = NULL;
   void *secret_ = NULL;
   size_t secretSize_;
 };

--- a/src/xxhash32.cc
+++ b/src/xxhash32.cc
@@ -4,11 +4,17 @@
 XXHash32::XXHash32(XXH32_hash_t seed)
     : env_(nullptr), wrapper_(nullptr), seed_(seed)
 {
-  XXH32_reset(&state_, seed);
+  state_ = XXH32_createState();
+  XXH32_reset(state_, seed);
 }
 
 XXHash32::~XXHash32()
 {
+  if (state_ != NULL)
+  {
+    XXH32_freeState(state_);
+    state_ = NULL;
+  }
   napi_delete_reference(env_, wrapper_);
 }
 
@@ -219,7 +225,7 @@ napi_value XXHash32::Update(napi_env env, napi_callback_info info)
   status = napi_unwrap(env, jsthis, reinterpret_cast<void **>(&hasher));
   assert(status == napi_ok);
 
-  XXH_errorcode hash_error = XXH32_update(&(hasher->state_),
+  XXH_errorcode hash_error = XXH32_update(hasher->state_,
                                           reinterpret_cast<void *>(data),
                                           buf_len);
   assert(hash_error == XXH_OK);
@@ -240,7 +246,7 @@ napi_value XXHash32::Digest(napi_env env, napi_callback_info info)
   assert(status == napi_ok);
 
   XXH32_canonical_t canonical_sum;
-  XXH32_hash_t sum = XXH32_digest(&(hasher->state_));
+  XXH32_hash_t sum = XXH32_digest(hasher->state_);
 
   // canonicalize the hash value
   XXH32_canonicalFromHash(&canonical_sum, sum);
@@ -266,7 +272,7 @@ napi_value XXHash32::Reset(napi_env env, napi_callback_info info)
   status = napi_unwrap(env, jsthis, reinterpret_cast<void **>(&hasher));
   assert(status == napi_ok);
 
-  assert(XXH32_reset(&(hasher->state_), hasher->seed_) == XXH_OK);
+  assert(XXH32_reset(hasher->state_, hasher->seed_) == XXH_OK);
 
   return nullptr;
 }

--- a/src/xxhash32.h
+++ b/src/xxhash32.h
@@ -32,7 +32,7 @@ private:
   napi_env env_;
   napi_ref wrapper_;
   XXH32_hash_t seed_;
-  XXH32_state_t state_;
+  XXH32_state_t* state_ = NULL;
 };
 
 #endif

--- a/src/xxhash64.cc
+++ b/src/xxhash64.cc
@@ -4,11 +4,17 @@
 XXHash64::XXHash64(XXH64_hash_t seed)
     : env_(nullptr), wrapper_(nullptr), seed_(seed)
 {
-  XXH64_reset(&state_, seed);
+  state_ = XXH64_createState();
+  XXH64_reset(state_, seed);
 }
 
 XXHash64::~XXHash64()
 {
+  if (state_ != NULL)
+  {
+    XXH64_freeState(state_);
+    state_ = NULL;
+  }
   napi_delete_reference(env_, wrapper_);
 }
 
@@ -228,7 +234,7 @@ napi_value XXHash64::Update(napi_env env, napi_callback_info info)
   status = napi_unwrap(env, jsthis, reinterpret_cast<void **>(&hasher));
   assert(status == napi_ok);
 
-  XXH_errorcode hash_error = XXH64_update(&(hasher->state_),
+  XXH_errorcode hash_error = XXH64_update(hasher->state_,
                                           reinterpret_cast<void *>(data),
                                           buf_len);
   assert(hash_error == XXH_OK);
@@ -249,7 +255,7 @@ napi_value XXHash64::Digest(napi_env env, napi_callback_info info)
   assert(status == napi_ok);
 
   XXH64_canonical_t canonical_sum;
-  XXH64_hash_t sum = XXH64_digest(&(hasher->state_));
+  XXH64_hash_t sum = XXH64_digest(hasher->state_);
 
   // canonicalize the hash value
   XXH64_canonicalFromHash(&canonical_sum, sum);
@@ -275,7 +281,7 @@ napi_value XXHash64::Reset(napi_env env, napi_callback_info info)
   status = napi_unwrap(env, jsthis, reinterpret_cast<void **>(&hasher));
   assert(status == napi_ok);
 
-  assert(XXH64_reset(&(hasher->state_), hasher->seed_) == XXH_OK);
+  assert(XXH64_reset(hasher->state_, hasher->seed_) == XXH_OK);
 
   return nullptr;
 }

--- a/src/xxhash64.h
+++ b/src/xxhash64.h
@@ -32,7 +32,7 @@ private:
   napi_env env_;
   napi_ref wrapper_;
   XXH64_hash_t seed_;
-  XXH64_state_t state_;
+  XXH64_state_t* state_ = NULL;
 };
 
 #endif

